doc/mgmt/SONiC_YANG_Model_Guidelines.md

```diff
@@ -0,0 +1,959 @@
+                   Both ip4 and ip6 address are string in IETF RFC 6020,
+                   so must statement can check based on : or ., family
+                   should be IPv4 or IPv6 according.
+                */
+                must "(contains(../ip-prefix, ':') and current()='IPv6') or
+                      (contains(../ip-prefix, '.') and current()='IPv4')";
+                type sonic-head:ip-family;
+            }
+```
+
+
+### 18. If a List object is needed in YANG model to bundle multiple entries from a Table in ABNF.json, but this LIST is not a valid entry in config data, then we must define such list as <TABLE_NAME>_LIST .
+
+For Example: Below entries in PORTCHANNEL_INTERFACE Table must be part of List Object in YANG model, because variable number of entries may be present in config data. But there is no explicit list in config data. To support this, a list object with name PORTCHANNEL_INTERFACE_LIST should be added in YANG model.
+#### ABNF:
+```
+"PORTCHANNEL_INTERFACE": {
+        "PortChannel01|10.0.0.56/31": {},
+        "PortChannel01|FC00::71/126": {},
+        "PortChannel02|10.0.0.58/31": {},
+        "PortChannel02|FC00::75/126": {}
+        ...
+    }
+```
+
+#### YANG
+```
+container PORTCHANNEL_INTERFACE {
+
+        description "PORTCHANNEL_INTERFACE part of config_db.json";
+
+        list PORTCHANNEL_INTERFACE_LIST {<<<<<
+        .....
+        .....
+        }
+}
+```
+
+### 19. In some cases it may be required to split an ABNF table into multiple YANG lists based on the data stored in the ABNF table. 
+
+Example : "INTERFACE" table stores VRF names to which an interface belongs, also it stores IP address of each interface. Hence it is needed to split them into two different YANG lists.
+
+#### ABNF
+```
+"INTERFACE" : {
+	"Ethernet1" : {
+		"vrf-name": "vrf1"
+	}
+	"Ethernet1|10.184.230.211/31": {
+	}
+}
+```
+#### YANG
+```
+......
+container sonic-interface {
+	container INTERFACE {
+		list INTERFACE_LIST {  // 1st list
+    			key ifname;
+			
+			leaf ifname {
+				type leafref { 
+				......
+				}
+			}
+			leaf vrf-name {
+				type leafref { 
+				......
+				}
+			}
+    			......
+		}
+
+		list INTERFACE_IPADDR_LIST { //2nd list
+   			key ifname, ip_addr;
+			
+  			leaf ifname {
+				type leafref { 
+				......
+				}
+			}
+  			leaf ip_addr {
+      				  type inet:ipv4-prefix;
+  			}
+			......
+		}
+	} 
+} 
+......
+```
+
+### 20. Add read-only nodes for state data using 'config false' statement. Define a separate top level container for state data. If state data is defined in other DB than CONFIG_DB, use extension 'sonic-ext:db-name' for defining the table present in other Redis DB. The default separator used in table key  is "|", if it is different, use 'sonic-ext:key-delim {separator};' YANG extension. This step applies when SONiC YANG is used as Northbound YANG.
+
+Example:
+
+#### YANG
+```
+container ACL_RULE {
+	list  ACL_RULE_LIST {
+	sonic-ext:db-name "APPL_DB"; //For example only
```

Comment:
ACL_RULE itself is created in CONFIG-DB only. If you intend to map the fields of container 'state' to a different table as source, please indicate it within the 'container state'.